### ISSUE TITLE
Task output does not seem to be cached (despite `cache_for` argument being present)

### ISSUE THREAD

**Issue by maxalbert** [PrefectHQ/prefect#1221]
_(I'm using prefect version `0.5.5`.)_

Here is a simple flow containing a single task that has the `cache_for` argument set.

I would expect this flow to take ~2s the first time it runs and then pretty much complete instantaneously on subsequent runs.
```python
from prefect import task, Flow
from datetime import timedelta
import time

@task(cache_for=timedelta(hours=1))
def return_the_answer():
    time.sleep(2)
    return 42

with Flow("answer") as flow:
    return_the_answer()

flow.run()
```

Here is the output from running in a Jupyter notebook using `%%time`, which confirms that the flow takes ~2s to complete.
```
[2019-07-10 20:45:09,227] INFO - prefect.FlowRunner | Beginning Flow run for 'answer'
[2019-07-10 20:45:09,229] INFO - prefect.FlowRunner | Starting flow run.
[2019-07-10 20:45:09,235] INFO - prefect.TaskRunner | Task 'return_the_answer': Starting task run...
[2019-07-10 20:45:11,240] INFO - prefect.TaskRunner | Task 'return_the_answer': finished task run for task with final state: 'Cached'
[2019-07-10 20:45:11,241] INFO - prefect.FlowRunner | Flow run SUCCESS: all reference tasks succeeded
CPU times: user 14.6 ms, sys: 4.24 ms, total: 18.8 ms
Wall time: 2.02 s
<Success: "All reference tasks succeeded.">
```

The strange thing is that subsequent runs take exactly the same amount of time but I'd expect the result to be cached so they should complete much more quickly.

It's quite possible that this is a misunderstanding on my part rather than a bug. Am I supplying the `cache_for` argument in the wrong place? (I couldn't find an example in the docs which applies it to tasks defined via the `@task` decorator.) Any clarifications would be appreciated. Many thanks!

**Comment by cicdw** [comment]
Hi @maxalbert - this is an excellent question.  There is some misunderstanding, but it is very subtle.

First off, you are providing the `cache_for` argument in the correct place - the `@task` decorator accepts all Task initialization keyword arguments.

The subtlety is that caching behaves a little differently when you are using only the open source engine.  Because the engine alone doesn't have a permanent place to store data, your cache only lives in memory for the duration of the `flow.run()` call.  This means that if you were to run this flow _on a schedule_ and call `flow.run()`, you will observe that caching is occurring in the way you expect.

On the Prefect Cloud platform, your cache will (essentially) live in the database, and so it persists across runs.  

I haven't found a satisfactory middle ground yet that both satisfies the expectations of open source users and is compatible with our API.  Since you want to use caching, let me ask you: where would you expect your cached data to live in this example, and how would you expect to be able to configure that location?

**Comment by maxalbert** [comment]
Thanks for the super quick reply and the clarification. 👍 

That's an excellent point about where the cache should live. I hadn't considered that of course if the workflow is defined in a script which is executed at different times then the different flow objects that are created during those executions need to be able to have a shared location to check the results from previous runs.

The context in which I ran into this issue was during prototyping, where I was defining and executing everything within the same interactive Python session - and thus calling `flow.run()` on flows which referenced the _same_ task object multiple times. Therefore my expectation was that the task would simply cache results in-memory and then re-use it while the Python session is still active (effectively emulating `functools.lru_cache()` or similar).

(The lengthy task in question was the first one in a sequence of a few tasks, and I was tweaking those subsequent tasks (and thus re-defining the flow which contained them) but I was keeping the first task fixed, so I expected the result to still be cached. I hope this makes sense, I'm happy to provide an example of my interactive session if required.)

Having this in-memory caching functionality would be quite useful during prototyping where you want to avoid having to wait for a long-running task every time you do a test run of the flow while you are tweaking some tasks in it.

Would this be something worth considering to support in Prefect? (And if so, would it be difficult to implement?)

**Comment by cicdw** [comment]
Ah that makes a lot of sense!  Yea I think there's room for some sort of global persistent dictionary that stores cached states in memory.  Let me try to prototype something and I'll ping you when I've got it working.

**Comment by maxalbert** [comment]
Awesome, thank you very much!

**Comment by cicdw** [comment]
@maxalbert I've implemented a version of this in #1226 - assuming others approve, once it's merged you should be good to go!  Note the new `cache_key` attribute on Tasks as well, which allows you to share caches amongst different tasks and even different flows.

Thanks for the issue, this was surprisingly easy to implement and I think will provide a large lift for other users as well!

**Comment by maxalbert** [comment]
That's great news! :)

I just tried it with your branch and can confirm it works as expected.

Kudos once again for the quick response and the solution, much appreciated!
